dspy/retrieve/MyScaleRM.py

```diff
@@ -0,0 +1,222 @@
+        self.metadata_columns = metadata_columns
+        self.vector_column = vector_column
+        self.k = k
+        self.openai_api_key = openai_api_key
+        self.model = openai_model
+        self.use_local_model = False
+
+        if local_embed_model:
+            self.setup_local_model(local_embed_model)
+        elif openai_api_key:
+            os.environ['OPENAI_API_KEY'] = self.openai_api_key
+
+    def setup_local_model(self, model_name: str):
+        """
+        Configures a local model for embedding generation, including model and tokenizer loading.
+
+        Args:
+            model_name: The name or path to the pre-trained model to load.
+
+        Raises:
+            ModuleNotFoundError: If necessary libraries (torch or transformers) are not installed.
+        """
+        try:
+            import torch
+            from transformers import AutoModel, AutoTokenizer
+        except ImportError as exc:
+            raise ModuleNotFoundError(
+                "You need to install PyTorch and Hugging Face's transformers library to use a local embedding model.",
+            ) from exc
+
+        self._local_embed_model = AutoModel.from_pretrained(model_name)
+        self._local_tokenizer = AutoTokenizer.from_pretrained(model_name)
+        self.use_local_model = True
+
+        if torch.cuda.is_available():
+            self.device = torch.device('cuda:0')
+        elif torch.backends.mps.is_available():
+            self.device = torch.device('mps')
+        else:
+            self.device = torch.device('cpu')
+
+        self._local_embed_model.to(self.device)
+
+    def get_embeddings(self, queries: List[str]) -> List[List[float]]:
+        """
+        Determines the appropriate source (OpenAI or local model) for embedding generation based on class configuration,
+        and retrieves embeddings for the provided queries.
+
+        Args:
+            queries: A list of text queries to generate embeddings for.
+
+        Returns:
+            A list of embeddings, each corresponding to a query in the input list.
+
+        Raises:
+            ValueError: If neither an OpenAI API key nor a local model has been configured.
+        """
+        if self.openai_api_key and self.model:
+            return self._get_embeddings_from_openai(queries)
+        elif self.use_local_model:
+            return self._get_embedding_from_local_model(queries)
+        else:
+            raise ValueError("No valid method for obtaining embeddings is configured.")
+
+    def _get_embeddings_from_openai(self, queries: List[str]) -> List[List[float]]:
+        """
+        Uses the OpenAI API to generate embeddings for a list of queries.
+
+        Args:
+            queries: A list of strings for which to generate embeddings.
+
+        Returns:
+            A list of lists, where each inner list contains the embedding of a query.
+        """
+        response = openai.embeddings.create(
```

Comment:
can these requests be cached? could you adapt from [DSPy's caching functionalities](https://github.com/stanfordnlp/dspy/blob/f2a96137288925be7a9ad46b5c40c1e0425bd531/dsp/modules/cache_utils.py#L9)?